Math: library: Added initial version of cordicsin

ShriramShastry opened the pull request:
Implement cordic sine algorithm with angle in the range [-pi/2, pi/2] having total harmonic distortion (THD) of (-170) dBc for SNR 

Signed-off-by: ShriramShastry <malladi.sastry@intel.com>


![image](https://user-images.githubusercontent.com/61411947/112634238-5ccee680-8e60-11eb-9aea-d94348252e5a.png)

**figure computation time per sample**

![image](https://user-images.githubusercontent.com/61411947/112634844-1928ac80-8e61-11eb-9fc1-cfac4656af37.png)

lgirdwood commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -105,16 +123,170 @@ inline int32_t drc_log_fixed(int32_t x)
+  bool negate;
+
+  th = thRadFxp;
+	 if (th > PIQ20OFFSET0) {											 /* >20bits */
```
The indentation looks off here, should be tab size 8 (but sometimes GH does not show this correctly)

cujomalainey replied:
I find if github is off usually that is a sign of tabs v spaces

marc-hb replied:
checkpatch reports many indentation issues.

lgirdwood reviewed (commented):
I can only comment on style, can @singalsu or @johnylin76 review for content. Thanks

cujomalainey commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -127,11 +299,11 @@ inline int32_t drc_asin_fixed(int32_t x)
 	 * If x <= 1/sqrt(2), then
-	 *   fpminimax(asin(x), [|1,3,5,7|], [|SG...|], [-1e-30;1/sqrt(2)], absolute)
-	 *   max err ~= 1.89936e-5
+	 *	  fpminimax(asin(x), [|1,3,5,7|], [|SG...|], [-1e-30;1/sqrt(2)], absolute)
```
just curious why we are changing white space down here?

singalsu commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -105,16 +123,170 @@ inline int32_t drc_log_fixed(int32_t x)
+	 }
+
+	 th <<= 9;															/* thRadFxp *= 512 */
+	 if ((th & PIQ30OFFSET0) != 0) {								/*	 NonZero 30bits */
```
Is just `if (th & PIQ30OFFSET0)` sufficient?

singalsu commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -105,16 +123,170 @@ inline int32_t drc_log_fixed(int32_t x)
+	 for (b_idx = 0; b_idx < 30; b_idx++) {							/*	 Iterate data with lookup table */
+		if (z < 0) {															/*	 negative data */
+		  th = z + inpLUT[b_idx];
+		  if ((th & PIQ30OFFSET0) != 0) {						/*	 30 bits */
```
This repeats many times. You could make an inline function with a descriptive function name.

cujomalainey replied:
Was reviewing earlier in meeting with @ShriramShastry 
I think we can combine ideas
```
static inline int32_t func_name(val, bit1, bit2)
{
    val = val & bit1;
    return th ? th | -bit1 : th & bit2;
}

singalsu commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -9,10 +9,28 @@
 #define q_multq(a, b, q) ((int32_t)Q_MULTSR_32X32((int64_t)(a), b, q, q, q))
 
+/* #define PIQ20OFFSET0	  Q_CONVERT_FLOAT_PRECISION(1.570796012878417968750000f, 20, 0.0)			*/
+/* #define PIQ21OFFSET0	  Q_CONVERT_FLOAT_PRECISION(1.570796489715576171875000f, 21, 0.0)			*/
```
Though my Octave thinks that PI/2 with 26 decimals is 1.57079632679489655799898173. Both of above differ from 7th decimal. Are they some magic numbers but not pi/2?

marc-hb replied:
> Both of above differ from 7th decimal.

`f` suffix means single-precision so most of these decimals are pointless.
```shell
python -c 'print(len(bin(10000009))-2)' # -2 to account for the '0b' prefix.
24
```

https://en.wikipedia.org/wiki/Single-precision_floating-point_format
>  The IEEE 754 standard specifies a binary32 as having:
> Sign bit: 1 bit
> Exponent width: 8 bits
> Significand precision: 24 bits (23 explicitly stored)
> This gives from 6 to 9 significant decimal digits precision.

ShriramShastry replied:
Thanks! Now single precision float is left out and current change is using double

ShriramShastry replied:
> Though my Octave thinks that PI/2 with 26 decimals is 1.57079632679489655799898173. Both of above differ from 7th decimal. Are they some magic numbers but not pi/2?

Just updated the PR , angle must converge for [-pi/2,pi/2].

singalsu commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -9,10 +9,28 @@
 
+/* #define PIQ20OFFSET0	  Q_CONVERT_FLOAT_PRECISION(1.570796012878417968750000f, 20, 0.0)			*/
+/* #define PIQ21OFFSET0	  Q_CONVERT_FLOAT_PRECISION(1.570796489715576171875000f, 21, 0.0)			*/
+/* #define PIQ30OFFSET0	  Q_CONVERT_FLOAT_PRECISION(1.000000000000000000000000,	30, 0.0)			*/
```
In the code it looks more like a bit mask so usage of BIT() macro might be the best way.

johnylin76 commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -105,16 +123,170 @@ inline int32_t drc_log_fixed(int32_t x)
+ * +--------------+---------+----------+----------+---------+
+ * |	  WordLength| 31		 |	 32		|12.20	  | 3.29		|
+ * |FractionLength| 20		 |	 29		|			  |			|
+ * +--------------+---------+----------+----------+---------+
```
Please re-arrange the comment for the table alignment

johnylin76 commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -9,10 +9,28 @@
 #include <sof/math/numbers.h>
 #include <sof/math/trig.h>
-
+#include <stdbool.h>
```
do we need bool? the library looks like to just use int for booleans.

johnylin76 commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -9,10 +9,28 @@
+#define PIQ22OFFSET0	  6588397						/* Q10.22 */
+#define PIQ28OFFSET4	  326016644						/* Q4.28	 */
+#define PIQ29OFFSET1	  1073741823					/* Q3.29	 */
+#define PIQ30OFFSET0	  1073741824					/* Q2.30	 */
```
What does PIQ30OFFSET0 mean?

johnylin76 commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -105,16 +123,170 @@ inline int32_t drc_log_fixed(int32_t x)
+		negate = false;
+	 }
+
+	 th <<= 9;															/* thRadFxp *= 512 */
```
Any hazard of overflow should be considered here?

johnylin76 commented:
Thanks for the commit.
In my opinion we may move this cordicsin function into src/math/, rather than src/audio/drc/drc_math_generic.c, because it is a general sine approximation.
Currently sin_fixed() is in src/math/trig.c, cordicsin may also live there or create its own .h/.c files.

paulstelian97 reviewed (changes requested):
Blocking on the whitespace changes. Tabs are 8-wide, not 4-wide. Please configure your editor properly and adjust that.

singalsu commented on src/include/sof/audio/format.h:
```diff
@@ -79,6 +79,13 @@
+/* Arithmetic shift right */
+#define ARITHMETICSHIFTRIGHT(u, n) ({	\
+	(u > 0) ?	\
+		(int32_t)((uint32_t)(((uint32_t)u) >> ((uint64_t)n))) :	\
```
Why is the shift right amount cast to 64 bit type?

paulstelian97 replied:
I guess the shifted value itself shall be cast to 64-bit? One of them needs to be 64-bit for the result to work in 64-bit domain.

lgirdwood replied:
This looks wrong and may generate inefficient code on 32bit devices. I would also change this to be a static inline function since the types are important here. e.g. (and if I understand the usage correctly)
```c
static inline uint32_t arith_shift_right(uint32_t u, uint32_t n)
{
    if (u > 0)
        return u >> n;
    else
        return abs(u) >> n;
}
```
We will also need an abs().

ShriramShastry replied:
@lgirdwood , Thanks for  your proposal. I think with your proposal- msb bit will be lost i.e. the  sign bit
For Example : 

```
 +---+----------+--------+---------+
 |n  |  u       |   y    |   y_liam|
 +---+----------+--------+---------+
 |4  | -50940067|-3183755|  3183754|
 +---+----------+--------+---------+

```

**In matlab -** 
bitsra(-50940067,4)
ans =
-3183754.1875

Test result with few sample data on signed bit

                                            =========================================

                                            =========================================
                                            ---Aithmetic shift right on signed bit---
                                            =========================================

                                            =========================================
           iter |   u-operand(i) |       nbits(i) |     y-outpt(i) |    Matlab-bsra |    c-out-fixed |      error-ppm |
           -----+----------------+----------------+----------------+----------------+----------------+----------------+
              1 |     -652032874 |              1 |  -3.260164e+08 |     -326016437 |     -326016437 |       0.000000 |
              2 |     -326016437 |              2 |      -81504110 |  -8.150411e+07 |      -81504110 |       0.000000 |
              3 |      -81504110 |              3 |  -1.018801e+07 |  -1.018801e+07 |      -10188014 |       0.000000 |
              4 |      -15918772 |              5 |  -4.974626e+05 |  -4.974616e+05 |        -497462 |       0.000000 |
              5 |       -7550352 |              8 |  -2.949456e+04 |  -2.949356e+04 |         -29494 |       0.000000 |
              6 |      -11744509 |              9 |  -2.293949e+04 |  -2.293849e+04 |         -22939 |       0.000000 |
              7 |      -13841530 |             10 |  -1.351812e+04 |  -1.351712e+04 |         -13518 |       0.000000 |
              8 |      -12793042 |             11 |  -6.247602e+03 |  -6.246603e+03 |          -6247 |       0.000000 |
              9 |      -13317292 |             12 |  -3.252292e+03 |  -3.251292e+03 |          -3252 |       0.000000 |
             10 |      -13055169 |             13 |  -1.594648e+03 |  -1.593649e+03 |          -1594 |       0.000000 |
             11 |      -13186231 |             14 |  -8.058236e+02 |  -8.048237e+02 |           -805 |       0.000000 |
             12 |      -13120700 |             15 |  -4.014120e+02 |  -4.004120e+02 |           -401 |       0.000000 |
             13 |      -13153465 |             16 |  -2.017059e+02 |  -2.007059e+02 |           -201 |       0.000000 |
             14 |      -13169847 |             17 |  -1.014780e+02 |  -1.004780e+02 |           -101 |       0.000000 |
             15 |      -13178038 |             18 |  -5.127022e+01 |  -5.027023e+01 |            -51 |       0.000000 |
             16 |      -13173943 |             19 |  -2.612730e+01 |  -2.512730e+01 |            -26 |       0.000000 |
             17 |      -13175990 |             20 |  -1.356560e+01 |  -1.256560e+01 |            -13 |       0.000000 |
             18 |      -13177013 |             21 |  -7.283289e+00 |  -6.283289e+00 |             -7 |       0.000000 |
             19 |      -13176502 |             22 |  -4.141523e+00 |  -3.141523e+00 |             -4 |       0.000000 |
             20 |      -13176757 |             23 |  -2.570792e+00 |  -1.570792e+00 |             -2 |       0.000000 |
             21 |      -13176884 |             24 |  -1.785403e+00 |  -7.854035e-01 |             -1 |       0.000001 |

singalsu commented on src/include/sof/audio/format.h:
```diff
@@ -79,6 +79,13 @@
 #define SATM_INT32(x) (((x) < INT32_MIN) ? INT32_MIN : (x))
 
+/* Arithmetic shift right */
+#define ARITHMETICSHIFTRIGHT(u, n) ({	\
```
How does this macro differ from C language shift ">>" for signed integer type? It's arithmetic shift.

ShriramShastry replied:
Yes, it's arithmetic shift

 Arithmetic Shift Right operates on both unsigned and signed inputs. If the input is unsigned, asr shifts zeros into the
 positions of bits that it shifts right. If the input is signed, asr shifts the MSB into the positions
 of bits that it shifts right.

right shifts on signed integers first casts the signed integers to unsigned integers, and then right shifts the unsigned integers, Using an arithmetic shift, the sign bit would be kept.

singalsu commented on src/audio/drc/drc_math_generic.c:
```diff
@@ -127,11 +127,11 @@ inline int32_t drc_asin_fixed(int32_t x)
+	 *	  fpminimax(asin(x), [|1,3,5,7|], [|SG...|], [1/sqrt(2);1], absolute)
+	 *	  max err ~= 3.085226e-2
 	 */
 	const int32_t TWO_OVER_PI = Q_CONVERT_FLOAT(0.63661977236758134f, qcl); /* 2/pi */
```
What is "qcl"? It should be a macro or literal, otherwise Q_CONVERT_FLOAT() produces float library C code and increases a lot code size.

Also I'd not use "f" for decimal number literals since by default without a type letter it would be double precision with better precision.

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,79 @@
+	xtmp = ARITHMETICSHIFTRIGHT(xn, (uint32_t)((int32_t)(b_idx + 1)));
+	ytmp = ARITHMETICSHIFTRIGHT(b_yn, (uint32_t)((int32_t)(b_idx + 1)));
+	}
+	cdc_sin_th = negate != 0 ? -b_yn : b_yn;
```
If negate could be +1 or -1 this would be a simple multiplication operation

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,79 @@
+		(th_rad_fxp < 0) ?
+			((th_rad_fxp += inpLUT[b_idx]), (xn += ytmp), (b_yn -= xtmp)) :
+			((th_rad_fxp -= inpLUT[b_idx]), (xn -= ytmp), (b_yn += xtmp));
+	xtmp = ARITHMETICSHIFTRIGHT(xn, (uint32_t)((int32_t)(b_idx + 1)));
```
Indentation is wrong here

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,79 @@
+	xtmp = CORDICSINE_LUT_Q29FL;
+	ytmp = 0;
+	for (b_idx = 0; b_idx < 31; b_idx++) {
+		(th_rad_fxp < 0) ?
```
Hard to read, would it be better readable with if statement?

lgirdwood commented on src/include/sof/audio/format.h:
```diff
@@ -3,8 +3,8 @@
- *         Liam Girdwood <liam.r.girdwood@linux.intel.com>
- *         Keyon Jie <yang.jie@linux.intel.com>
+ *			  Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ *			  Keyon Jie <yang.jie@linux.intel.com>
```
@ShriramShastry I think your editor was trying to be too helpful and converted some spaces to tabs and vice versa. Probably worth switching it off. Fwiw, if you use ```git gui``` you can manually pick the lines you want to commit and then discard the rest that are not needed (useful when your editor is being too helpful!).

cujomalainey commented on src/math/trig.c:
```diff
@@ -3,574 +3,79 @@
 
-	return (int32_t)sine;
+	negate = (th_rad_fxp > CORDICSINE_PIOVERTWO_Q28FL) ?
+	((th_rad_fxp - CORDICSINE_PIOVERTWO_Q29FL) <= CORDICSINE_PIOVERTWO_Q28FL) ?
```
this is not easy to read unfortunately. I wore recommend we avoid using statements in ternaries and stick to expressions,

paulstelian97 replied:
Also ternary operator isn't faster than simple if-then-else structures, so in this specific situation where the condition and values are highly intricate it's completely useless and leading to lack of readability with **no** gain whatsoever.

cujomalainey commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
-//         Keyon Jie <yang.jie@linux.intel.com>
-
+// Liam Girdwood <liam.r.girdwood@linux.intel.com>
+// Keyon Jie <yang.jie@linux.intel.com>
```
not sure why this is a diff but regardless this fixes the indentation

paulstelian97 replied:
Actually before the authors were all aligned. Shouldn't we have an "Author: " thingy on each of the authors with SPDX actually?

cujomalainey replied:
Oh I see what is going on, I missed the Author: prefix when I edited the whitespace. Thanks for pointing that out.

ShriramShastry commented:
From previous check-in
```

1. Re written  ternary to if else for better readable code. 
2. Changed arithmetic shift right typecast uint64_t for raise to power to unit32_t.
3. Description for Arithmetic shift right :  Arithmetic Shift Right operates on both unsigned and signed inputs. If the input is 
unsigned, asr shifts zeros into the positions of bits that it shifts right. If the input is signed, asr shifts the MSB into the positions
of bits that it shifts right.
4. Corrected  editor indentation

```

paulstelian97 commented:
> ```
> 3. Description for Arithmetic shift right :  Arithmetic Shift Right operates on both unsigned and signed inputs. If the input is 
> unsigned, asr shifts zeros into the positions of bits that it shifts right. If the input is signed, asr shifts the MSB into the positions
> of bits that it shifts right.
> ```

That's what the C built in shift operator does, as long as your integers are properly sized. Sure, for 24-bit it's a bit annoying because there's no native 24-bit signed integer type but otherwise the built in operator actually works that exact way.

paulstelian97 commented on src/include/sof/audio/format.h:
```diff
@@ -79,6 +79,13 @@
+	(u > 0) ?	\
+		(int32_t)((uint32_t)(((uint32_t)u) >> ((uint32_t)n))) :	\
+		(-((int32_t)((uint32_t)(((uint32_t)((int32_t)(-1 - u))) >> ((uint32_t)n))))) - 1; \
+})
```
Honestly, holy fuck, this code looks like crap and there's no real need for it in any case.

(int)0x80000000 >> 31 == 0xFFFFFFFF

cujomalainey replied:
Agreed its not pretty, but we all have different backgrounds. Some of us more mathematical and some of us more software, but that is why we work as a team to get code in. @ShriramShastry I am not sure what the point of this macro is, can you detail why there is a need for so many casts?

lgirdwood commented on src/include/sof/audio/format.h:
```diff
@@ -79,6 +79,13 @@
+	(u > 0) ?	\
+		(int32_t)((uint32_t)(((uint32_t)u) >> ((uint32_t)n))) :	\
+		(-((int32_t)((uint32_t)(((uint32_t)((int32_t)(-1 - u))) >> ((uint32_t)n))))) - 1; \
+})
```
Please see my previous comments. This needs to be a static inline function which is type safe.

lgirdwood commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
+   /* 1686629713, deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q30FL	= Q_CONVERT_FLOAT(1.570796326734126, 30);
+   /* 652032874 , deg = 69.586061*/
+	const int32_t CORDICSINE_LUT_Q29FL =	 Q_CONVERT_FLOAT(1.214505869895220, 29);
```
best to have a newline here after the last variable declaration. Btw, the indentation looks a bit odd above ?

paulstelian97 replied:
Indentation looks odd due to diff/Github.

lgirdwood reviewed (changes requested):
Mostly minor updates needed.

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
+			negate = 1;
+		} else {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q30FL;
+			negate = 0;
```
This to negate = 1, maybe also rename to sign or something

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
 
-	if (idx > 2 * SINE_NQUART)
-		s = -sine_table[i1];
+	if (negate)
```
And here cdc_sin = sign * b_yn;

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
+			xn -= ytmp;
+			b_yn += xtmp;
+		}
+	xtmp = ARITHMETIC_SHIFT_RIGHT(xn, (uint32_t)((int32_t)(b_idx + 1)));
```
`xtmp = xn >> (b_idx +1 );` and same for ytmp.

singalsu reviewed (changes requested):
A lot of improvement, can you please check if my proposals make sense.

singalsu commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
+		cdc_sin_th = b_yn;
 
-	return (int32_t)sine;
+	return (int32_t)cdc_sin_th;
```
I think the type is already int32_t so the typecast is unnecessary.

cujomalainey commented on src/math/trig.c:
```diff
@@ -3,574 +3,100 @@
+	int32_t xtmp;
+	int32_t ytmp;
+	int32_t cdc_sin_th;
+	int32_t negate;
```
I think int would be better given we know xtensa doesn't like small types.

lgirdwood commented:
Please use "main" as PR target.
